src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -3619,15 +3620,38 @@ void MacroAssembler::encode_klass_not_null(Register dst, Register src) {
   bind(ok);
 #endif
 
+  // Do Klass pointers always fit into 32bit?
+  const MetaWord* ccs_end = Metaspace::class_space_end();
+  assert(ccs_end != NULL, "Sanity");
+  const bool klass_pointers_are_32_bit = ((uintptr_t)ccs_end <= 0xFFFFFFFF);
+
+#ifdef ASSERT
+  if (klass_pointers_are_32_bit) {
+    // Assert that Klass pointers are < 32bit
+    Label ok2;
+    z_tmhl(current, 0xFFFF);
+    z_brc(Assembler::bcondAllZero, ok2);
+    z_illtrap(0xee);
+    z_illtrap(0xee);
+    bind(ok2);
+    Label ok3;
+    z_tmhh(current, 0xFFFF);
+    z_brc(Assembler::bcondAllZero, ok3);
+    z_illtrap(0xee);
+    z_illtrap(0xee);
+    bind(ok3);
+  }
+#endif
+
```

Comment:
> I do not like the cross-dependency to metaspace.hpp just for the sake of checking an artificial restriction on Klass pointers.

It is not just for the assertion, it is for limiting the 32bit add to situations where we know Klass pointers cannot exceed 32bit. That was the main reason. As I wrote, I was not sure about the assertion myself and am happy to drop it.

> And by the way, you could do the check with one test:
> 
> ```
>   z_oihf(current, 0);
>   z_brc(Assembler::bcondZero, ok);
> ```
> 
> z_oihf() does modify the contents of register current, but it writes back the same value.

Thank you. Unfortunately, information about z assembly was hard to come by. The only public information I found had hardly more than the instruction names, the rest was trial and error.

Reply:
I admit. To find System z information, you need to know the "magic keywords" to search for. In this case, it would be "Principles of Operation". The third or so Google hit would lead you to the System z architecture document. With 2000+ pages to read, you would be lost anyway. :-)